// antsibull-docs/dist/cjs/index.js
"use strict";
/*
  Simplified BSD License (see LICENSES/BSD-2-Clause.txt or https://opensource.org/licenses/BSD-2-Clause)
  SPDX-FileCopyrightText: Ansible Project
  SPDX-License-Identifier: BSD-2-Clause
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toRST = exports.quoteRST = exports.toMD = exports.quoteMD = exports.toHTML = exports.quoteHTML = exports.toAnsibleDocText = exports.dom = exports.parse = void 0;
// Parser
var parser_1 = require("./parser");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return parser_1.parse; } });
exports.dom = __importStar(require("./dom"));
// Output
var ansible_doc_text_1 = require("./ansible-doc-text");
Object.defineProperty(exports, "toAnsibleDocText", { enumerable: true, get: function () { return ansible_doc_text_1.toAnsibleDocText; } });
var html_1 = require("./html");
Object.defineProperty(exports, "quoteHTML", { enumerable: true, get: function () { return html_1.quoteHTML; } });
Object.defineProperty(exports, "toHTML", { enumerable: true, get: function () { return html_1.toHTML; } });
var md_1 = require("./md");
Object.defineProperty(exports, "quoteMD", { enumerable: true, get: function () { return md_1.quoteMD; } });
Object.defineProperty(exports, "toMD", { enumerable: true, get: function () { return md_1.toMD; } });
var rst_1 = require("./rst");
Object.defineProperty(exports, "quoteRST", { enumerable: true, get: function () { return rst_1.quoteRST; } });
Object.defineProperty(exports, "toRST", { enumerable: true, get: function () { return rst_1.toRST; } });
